### PR TITLE
Functions to plot spatial distributions of site selection frequencies

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
Adds function to viz extension which plots the spatial distribution of location selection frequency. PR #431 must be merged to main first.

Example usage is added to the viz documentation.

Color map specification is also added to the inputs of `ADRIA.viz.map()` to allow different colour maps to be specified in spatial plots. Default is set to `:GrayC`.

### CODE REVIEW

**Comment by ConnectedSystems** [comment]
This PR depends on changes in #449 which should be reviewed/considered first.

@Rosejoycrocker FYI you have to signpost such dependencies clearly.

On the submitted changes, I'm uncomfortable with making `color_map` an explicit and separate argument. To me, it seems this should be something handled by `opts` or similar.

**Comment by Rosejoycrocker** [comment]
> This PR depends on changes in #449 which should be reviewed/considered first.
> 
> @Rosejoycrocker FYI you have to signpost such dependencies clearly.
> 
> On the proposed changes, I'm uncomfortable with making `color_map` an explicit and separate argument. To me, it seems this should be something handled by `opts` or similar.

Hey @ConnectedSystems, I put a note in the description saying it is related to PR #431 - is that what you mean by signposting dependencies or something different?

I can make color_map a part of opts

**Comment by ConnectedSystems** [comment]
> Hey @ConnectedSystems, I put a note in the description saying it is related to PR https://github.com/open-AIMS/ADRIA.jl/pull/431 - is that what you mean by signposting dependencies or something different?

Yes, you said it's related but "related" doesn't indicate that the previous PR must be merged in before this one.

Once we have #431 merged in, you can rebase to clean your commit timeline. Otherwise we'll have duplicates.

**Comment by Rosejoycrocker** [comment]
> > Hey @ConnectedSystems, I put a note in the description saying it is related to PR #431 - is that what you mean by signposting dependencies or something different?
> 
> Yes, you said it's related but "related" doesn't indicate that the previous PR must be merged in before this one.
> 
> Once we have #431 merged in, you can rebase to clean your commit timeline. Otherwise we'll have duplicates.

Ok cool, thanks for clarifing

**Comment by Rosejoycrocker** [comment]
Hey @ConnectedSystems color_map is now added to opts, so ready for review. Should the documentastion for `series_opts` here: https://github.com/open-AIMS/ADRIA.jl/blob/354543eef9db02972bd01315108e0cfcaa2b0082/ext/AvizExt/viz/spatial.jl#L129
be changes to `fig_opts`?

**Comment by Zapiano** [comment]
@Rosejoycrocker just checking, is this ready for review?

**Comment by Rosejoycrocker** [comment]
> @Rosejoycrocker just checking, is this ready for review?

Thanks for checking @Zapiano . Due to some changes I'm waiting to discuss some details with @ConnectedSystems before it's ready :) I'll comment again when it is.

**Comment by Rosejoycrocker** [comment]
Example of plotting selection frequencies for Moore for ranks 1-3:

ranks_plot_ex

**Comment by ConnectedSystems** [comment]
If I'm interpreting the plot above correctly, there's no spread in the selection frequency, is that correct?

**Comment by Rosejoycrocker** [comment]
There is some spread but it's hard to see because I zoomed in on the part with the highest values. It's also a set of only ~200 runs so I'd expect more spread for a larger set

**Comment by ConnectedSystems** [comment]
Could I get a MWE for this? Something that shows how it could be used?

You say in the initial post that you've added one to the documentation but I don't see changes to any doc files.

**Comment by Rosejoycrocker** [comment]
> Could I get a MWE for this? Something that shows how it could be used?
> 
> You say in the initial post that you've added one to the documentation but I don't see changes to any doc files.

Sorry I missed this comment, I'll add one now. I can't remember what I meant when I said that

File: assets/imgs/example_n_selected_locations_spatial.png
Comment (ConnectedSystems) [inline assets/imgs/example_n_selected_locations_spatial.png:1]: I dislike having `example` / `ex` / `eg` in the filename. A better organisation approach would be to have images in separate directories for each page.

e.g., `assets/imgs/analysis/selected_locations_map.png`

I will raise a separate issue for this to be tackled.

File: docs/src/usage/analysis.md [hunk 1]
```diff
@@ -157,6 +157,38 @@ save("aviz_scenario.png", tf)  # save the figure to a file
...
+
+# Calculate frequencies with which each site was selected at each rank
+rank_freq = ranks_to_frequencies(
+    rs.ranks[intervention=1];
+    agg_func=x -> dropdims(sum(x; dims=:timesteps); dims=:timesteps),
+)
+
+# Plot 1st rank frequencies as a colormap
+tf = Figure(; resolution=(1200, 800))  # resolution in pixels
+ADRIA.viz.ranks_to_frequencies!(tf[1, 1], rs, rank_freq, 1)
```
Comment (ConnectedSystems) [inline docs/src/usage/analysis.md:174]: 1. Use the `fig` option to set resolution OR
2. Define a default resolution

File: docs/src/usage/analysis.md [hunk 2]
```diff
@@ -157,6 +157,38 @@ save("aviz_scenario.png", tf)  # save the figure to a file
...
+save("single_rank_plot_ex.png", tf)
+
+```
+
+![PAWN sensitivity plots](/ADRIA.jl/dev/assets/imgs/single_rank_plot_ex.png?raw=true "Rank frequency plot 1 rank")
+
+```julia
+# Plot 1st, 2nd and 3rd rank frequencies as an overlayed colormap
+tf = Figure(; resolution=(1200, 800))  # resolution in pixels
+ADRIA.viz.ranks_to_frequencies!(tf[1, 1], rs, rank_freq, [1, 2, 3])
```
Comment (ConnectedSystems) [inline docs/src/usage/analysis.md:185]: Same comment here - user should not have to create a Figure themselves in the first instance.

File: docs/src/usage/analysis.md [hunk 3]
```diff
@@ -157,6 +157,38 @@ save("aviz_scenario.png", tf)  # save the figure to a file
...
+rank_freq = ranks_to_frequencies(
+    rs.ranks[intervention=1];
+    agg_func=x -> dropdims(sum(x; dims=:timesteps); dims=:timesteps),
+)
+
+# Plot 1st rank frequencies as a colormap
+tf = Figure(; resolution=(1200, 800))  # resolution in pixels
+ADRIA.viz.ranks_to_frequencies!(tf[1, 1], rs, rank_freq, 1)
+
+save("single_rank_plot_ex.png", tf)
```
Comment (ConnectedSystems) [inline docs/src/usage/analysis.md:176]: Please get rid of the `ex` suffix in the filename.

File: ext/AvizExt/theme.jl [hunk 4]
```diff
@@ -15,7 +15,7 @@ const COLORS::Dict{Symbol,Symbol} = Dict(
 )
 
 function colors(
-    scen_groups::Dict{<:Any,BitVector}
+    scen_groups::Dict{Symbol,BitVector},
```
Comment (ConnectedSystems) [inline ext/AvizExt/theme.jl:18]: Why the change to `Symbol` here?

File: ext/AvizExt/viz/location_selection.jl [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    )
+
+    return ADRIA.viz.map!(
+        g,
+        rs,
+        AxisKeys.keyless(NamedDims.unname(frequencies[ranks=rank_id]));
+        opts=opts,
+        axis_opts=axis_opts,
+    )
+
```
Comment (ConnectedSystems) [inline ext/AvizExt/viz/location_selection.jl:110]: remove extra newline

File: ext/AvizExt/viz/location_selection.jl [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    rank_groups::Dict{Symbol,BitVector}, alpha_vals::Dict{Symbol,Float64}
+)
+    rank_colors = colors(rank_groups, alpha_vals)
+    rank_ids = keys(rank_groups)
+    return Dict(
+        rank_grp =>
+            [RGBA{Float32}(1.0, 1.0, 1.0, 0.1), rank_colors[rank_grp]] for
+        rank_grp in rank_ids
+    )
+
```
Comment (ConnectedSystems) [inline ext/AvizExt/viz/location_selection.jl:154]: Remove extra new line

File: ext/AvizExt/viz/location_selection.jl [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+end
```
Comment (ConnectedSystems) [inline ext/AvizExt/viz/location_selection.jl:155]: Missing new line at end of file?

File: ext/AvizExt/viz/spatial.jl [hunk 8]
```diff
@@ -8,26 +8,37 @@ Base.getindex(fc::AbstractFeatureCollection, i::Vector) = features(fc)[i]
...
 
 # Arguments
-- `f` : GLMakie figure to create plot in
-- `geodata` : FeatureCollection, Geospatial data to display
-- `data` : Values to use for choropleth
-- `highlight` : Stroke colors for each location
+- `f` : GLMakie figure to create plot in.
+- `geodata` : FeatureCollection, Geospatial data to display.
+- `data` : Values to use for choropleth.
+- `highlight` : Stroke colors for each location.
```
Comment (ConnectedSystems) [inline ext/AvizExt/viz/spatial.jl:20]: I don't think full stops are needed?

File: ext/AvizExt/viz/spatial.jl [hunk 9]
```diff
@@ -43,9 +54,9 @@ function create_map!(f::Union{GridLayout,GridPosition}, geodata::GeoMakie.GeoJSO
         dest="+proj=latlong +datum=WGS84",
         axis_opts...
     )
-    spatial.xticklabelsize = 12
-    spatial.yticklabelsize = 12
-
+    spatial.xticklabelsize = 14
+    spatial.yticklabelsize = 14
+
```
Comment (ConnectedSystems) [inline ext/AvizExt/viz/spatial.jl:59]: Superfluous spaces?

File: docs/src/usage/analysis.md [hunk 10]
```diff
@@ -157,6 +157,38 @@ save("aviz_scenario.png", tf)  # save the figure to a file
...
+
+Plot spatial colormaps of site selection frequencies and other available site selection metrics.
+
+```julia
+
+# Calculate frequencies with which each site was selected at each rank
+rank_freq = ranks_to_frequencies(
+    rs.ranks[intervention=1];
+    agg_func=x -> dropdims(sum(x; dims=:timesteps); dims=:timesteps),
+)
```
Comment (ConnectedSystems) [inline docs/src/usage/analysis.md:170]: I thought we agreed to use named dimensions, e.g., `intervention=:seed` - or is this in a separate PR?

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Some minor changes please

File: docs/src/usage/analysis.md (hunk 10 above)
Comment (Rosejoycrocker) [reply docs/src/usage/analysis.md:170]: Ideally yes but the ranks storage in ResultSet is created to have interventions named 1 and 2. I changed it in standalone site selection to be consistent with this. I imagine this will change when we move to YAXarrays (?), but for now  the dimension names are created automatically: https://github.com/open-AIMS/ADRIA.jl/blob/21349226ad68a0d00b8f5fea6adbf51efd97f8be/src/io/ResultSet.jl#L75-L79

File: ext/AvizExt/theme.jl (hunk 4 above)
Comment (Rosejoycrocker) [reply ext/AvizExt/theme.jl:18]: Not sure might have been testing something and forgot about it, I'll revert

File: docs/src/usage/analysis.md (hunk 10 above)
Comment (ConnectedSystems) [reply docs/src/usage/analysis.md:170]: Yes, sounds reasonable.

File: ext/AvizExt/theme.jl (hunk 4 above)
Comment (Zapiano) [reply ext/AvizExt/theme.jl]: Actually I think probably it makes more sense to change to Symbol, or there is something wrong here (I don't know this PR is the right place for that, tough). 

The return type of this function is a Dictionary when the keys are only allowed to be Symbols and they come from the keys of `scen_groups`, so this will raise an error if `scen_groups` has other type as keys... (right?)

Comment (Rosejoycrocker) [reply ext/AvizExt/theme.jl]: Yes @Zapiano that's right. I can't think of a case where we'd want non-symbol keys but we'd probably want to check first

Comment (Zapiano) [reply ext/AvizExt/theme.jl]: I'm going to work on Issue https://github.com/open-AIMS/ADRIA.jl/issues/549 now so I can take a look and change it in my PR, if that's ok with you.

Comment (Rosejoycrocker) [reply ext/AvizExt/theme.jl]: Thanks @Zapiano, sounds good :)

**Comment by Rosejoycrocker** [comment]
I think I've addressed your comments @ConnectedSystems

File: docs/src/usage/analysis.md [hunk 11]
```diff
@@ -157,6 +157,36 @@ save("aviz_scenario.png", tf)  # save the figure to a file
...
+![PAWN sensitivity plots](/ADRIA.jl/dev/assets/imgs/single_rank_plot.png?raw=true "Rank frequency plot 1 rank")
+
+```julia
+# Plot 1st, 2nd and 3rd rank frequencies as an overlayed colormap
+ADRIA.viz.ranks_to_frequencies(rs, rank_freq, [1, 2, 3]; fig_opts=Dict(:resolution=>(1200, 800)))
+
+save("ranks_plot.png", tf)
+
+```
+![PAWN sensitivity plots](/ADRIA.jl/dev/assets/imgs/ranks_plot.png?raw=true "Rank frequency plot 3 ranks")
```
Comment (ConnectedSystems) [inline docs/src/usage/analysis.md:188]: Sorry, why is the link text related to PAWN analysis?

Comment (Rosejoycrocker) [reply docs/src/usage/analysis.md:188]: Thanks for spotting, I copied and pasted the format and forgot to change the name

File: src/viz/viz.jl [hunk 12]
```diff
@@ -45,6 +45,10 @@ function clustered_scenarios!() end
 function rules_scatter() end
 function rules_scatter!() end
 
+# Location selection Plots
```
Comment (ConnectedSystems) [inline src/viz/viz.jl:48]:
```suggestion
# Location selection plots
```

**Comment by ConnectedSystems** [comment]
Tests error on this branch - looks like the model spec is missing details.

Did you rebase on top of the changes done recently where you included the missing default bounds for all components?

**Comment by Rosejoycrocker** [comment]
> Tests error on this branch - looks like the model spec is missing details.
> 
> Did you rebase on top of the changes done recently where you included the missing default bounds for all components?

I just rebased and all tests are passing on my end now :)

**Review by ConnectedSystems — APPROVED** [review]
